toolbox/HBTReader.py

```diff
@@ -1,279 +1,686 @@
+                not host subhalos
+        """
+        with self.OpenFile(isnap) as subfile:
+            trackIds = subfile['Membership/GroupedTrackIds'][HostHaloId]
+
+        try:
+            result = np.hstack(
+                [self.GetSub(trackId, isnap) for trackId in trackIds])
+        except ValueError:
+            result = np.array([])
+
+        return result
+
+    def GetMergerTree(self, HostHaloId, isnap=-1, file=None):
+        """Builds a FoF merger tree starting at a host halo ID.
+
+        Prints a graph which can be consumed by [Dot](graphviz.org] software.
+
+        Arguments:
+            HostHaloId (int): a halo at which to root a tree
+            isnap (int): (default = -1) snapshot number
+            file (File): (default=None) if not ``None``, Dot diagram will be generated
+                on the fly
+
+        Returns:
+            (list): multiply embedded list with a tree, i.e. ``[1, [2, 3]]``
+        """
+
+        progenitors = self.GetHostProgenitors(HostHaloId, isnap, file)
+
+        loggging.debug(
+            "Halo %d at %d with %d progenitor(s)" %
+            (HostHaloId, isnap, progenitors.size))
+
+        # TODO: print TrackId as graph edge label
+        if file is not None:
+            file.write("\t\"%d_%d\" [label=\"%d, %d\"]\n"%\
+                (isnap, HostHaloId, isnap, HostHaloId))
+            for progenitor in progenitors:
+                file.write("\t\"%d_%d\" -- \"%d_%d\"\n" %
+                           (isnap, HostHaloId, isnap - 1, progenitor))
+
+        return [(HostHaloId, isnap), [] if progenitors.size == 0 else [
+            self.GetMergerTree(progenitor, isnap - 1, file)
+            for progenitor in progenitors
+        ]]
+
+    def GetHostProgenitors(self, HostHaloId, isnap=-1):
+        """Returns progenitors of a given host halo.
+
+        Given a HostHaloId and a snapshot number, it finds all subhaloes within
+        the host, follows their track ids to the previous snapshot, checks if
+        they are the main / central subhalo (by ``Rank`` values), finds those
+        subhaloes' hosts and keeps the unique IDs.  Filtering by ``Rank`` is
+        necessary to remove fly-by and re-merger events.
+
+        Arguments:
+            HostHaloId (int): host halo id
+            isnap (int): (default=-1) snapshot number
+
+        Returns:
+            (numpy.ndarray): array of host halo ids of progenitors
+        """
+
+        try:
+            subhaloes = self.GetSubsOfHost(HostHaloId, isnap=isnap)['TrackId']
+            progenitors = np.unique([
+                self.GetSub(subhalo, isnap - 1)[0]['HostHaloId']
+                for subhalo in subhaloes
+                if self.GetSub(subhalo, isnap - 1)[0]['Rank'] == 0
+            ])
+        except:
+            progenitors = np.array([])
+
+        return progenitors
+
+    def GetCollapsedMassHistory(self, HostHaloId, isnap=-1, f=0.01):
+        """Calculates a CMH, starting at a FOF group.
```

Comment:
this function appear to be of special interest only. I prefer to keep it in the user's own module.